types/index.d.ts

```diff
@@ -16,10 +16,34 @@ export class Kafka {
 
 export type BrokersFunction = () => string[] | Promise<string[]>
 
+type SaslAuthenticationRequest = {
+  encode: () => Buffer | Promise<Buffer>
+}
+type SaslAuthenticationResponse<ParseResult> = {
+  decode: (rawResponse: Buffer) => Buffer | Promise<Buffer>
+  parse: (data: Buffer) => ParseResult
+}
+
+export type Authenticator = {
+  authenticate: () => Promise<void>
+}
+
+export type Mechanism = {
+  mechanism: string
+  authenticationProvider: (
+    connection: { host: string; port: number },
```

Comment:
I'm wondering if this field should just be flattened and host and port passed in directly.

Reply:
Probably makes sense. I think the built-in authenticators receive the entire `connection` object because they also want access to the `SASLOptions` for credentials and such, but that's not necessary for the custom authenticators.

Reply:
Thinking more about it, do you even need the host and the port? The authentication is done through `saslAuthenticate`, not by some bespoke connection to the broker anyway. I think the host and port are literally only used for logging in the existing authenticators. That does not seem necessary. The logger will log which broker it's communicating with when the SASL requests are made anyway.

Reply:
The AWS MSK IAM auth uses the host as part of generating the request object (as part of the generated signature) so because I don't know what other mechanisms will need I'll leave in the host/port fields but I'll pull them to the top level.